spring-ai-core/src/main/java/org/springframework/ai/vectorstore/observation/AbstractObservationVectorStore.java

```diff
@@ -48,45 +51,65 @@ public AbstractObservationVectorStore(ObservationRegistry observationRegistry,
 	public void add(List<Document> documents) {
 
-		VectorStoreObservationContext observationContext = this
-			.createObservationContextBuilder(VectorStoreObservationContext.Operation.ADD.value())
-			.build();
+		Supplier<VectorStoreObservationContext> observationContext = observationContextSupplier(
+				VectorStoreObservationContext.Operation.ADD);
 
 		VectorStoreObservationDocumentation.AI_VECTOR_STORE
-			.observation(this.customObservationConvention, DEFAULT_OBSERVATION_CONVENTION, () -> observationContext,
+			.observation(this.customObservationConvention, DEFAULT_OBSERVATION_CONVENTION, observationContext,
 					this.observationRegistry)
-			.observe(() -> this.doAdd(documents));
+			.observe(() -> doAdd(documents));
 	}
 
 	@Override
+	@SuppressWarnings("all")
 	public Optional<Boolean> delete(List<String> deleteDocIds) {
 
-		VectorStoreObservationContext observationContext = this
-			.createObservationContextBuilder(VectorStoreObservationContext.Operation.DELETE.value())
-			.build();
+		Supplier<VectorStoreObservationContext> observationContext = observationContextSupplier(
+				VectorStoreObservationContext.Operation.DELETE);
 
 		return VectorStoreObservationDocumentation.AI_VECTOR_STORE
-			.observation(this.customObservationConvention, DEFAULT_OBSERVATION_CONVENTION, () -> observationContext,
+			.observation(this.customObservationConvention, DEFAULT_OBSERVATION_CONVENTION, observationContext,
 					this.observationRegistry)
-			.observe(() -> this.doDelete(deleteDocIds));
+			.observe(() -> doDelete(deleteDocIds));
 	}
 
 	@Override
+	@SuppressWarnings("all")
 	public List<Document> similaritySearch(SearchRequest request) {
 
-		VectorStoreObservationContext searchObservationContext = this
-			.createObservationContextBuilder(VectorStoreObservationContext.Operation.QUERY.value())
-			.withQueryRequest(request)
-			.build();
+		Supplier<VectorStoreObservationContext> observationContext = observationContextSupplier(
+				VectorStoreObservationContext.Operation.QUERY, builder -> builder.withQueryRequest(request));
 
-		return VectorStoreObservationDocumentation.AI_VECTOR_STORE
-			.observation(this.customObservationConvention, DEFAULT_OBSERVATION_CONVENTION,
-					() -> searchObservationContext, this.observationRegistry)
-			.observe(() -> {
-				var documents = this.doSimilaritySearch(request);
-				searchObservationContext.setQueryResponse(documents);
-				return documents;
-			});
+		Observation observation = VectorStoreObservationDocumentation.AI_VECTOR_STORE.observation(
+				this.customObservationConvention, DEFAULT_OBSERVATION_CONVENTION, observationContext,
+				this.observationRegistry);
+
+		return observation.observe(() -> {
+			var documents = doSimilaritySearch(request);
+			getObservationContext(observation).ifPresent(context -> context.setQueryResponse(documents));
+			return documents;
+		});
+	}
+
+	private Optional<VectorStoreObservationContext> getObservationContext(@Nullable Observation observation) {
```

Comment:
`ChatModelObservationSupport` can be parameterized so it will work for multiple types, This might not be needed.

Reply:
I thought about that, too. In the end. I keep Chat and Vector separate for now, the only 2 use cases I know of.